### PR TITLE
search: use method to reduce nested closure

### PR DESCRIPTION
**Pull request by rvantonder**
@camdencheek we mentioned before about reducing closures in the `structural.go` file I was doing. I realized we can reduce this by using methods. Putting in draft, want to know what do you think of this? I do think it's cleaner and more "Go-like"

This made me also have a semi-epiphany to finally answer the question I've been dragging along for months re: "why would I really want to use methods in Go over functions?" (Somewhat facetious answer: "Go syntax makes it very awkward to nest functions or define partial functions. But, by adding a level of indirection and complicating your data type to an object + methods, you can remove the need for such awkward syntax!"). 

Addendum: I also realized in the meantime that the object + method approach works well at a local level to make things simpler. But! It breaks down when you spread things across packages (it's awkward to define methods on a type that is defined in a different package--here, functions don't have that issue, it's easy to pass arguments of types defined in different packages.)

### CODE REVIEW

File: internal/search/unindexed/structural.go [hunk 1]
```diff
@@ -42,29 +42,24 @@ func (UnindexedList) IsIndexed() bool {
...
-// structuralSearchJob creates a composable function for running structural
-// search. It unrolls the context parameter so that multiple jobs can be
-// parameterized by the errgroup context.
-func structuralSearchJob(args *search.TextParameters, stream streaming.Sender, repoData repoData) withContext {
-	return func(ctx context.Context) searchJob {
-		return func() error {
-			return callSearcherOverRepos(ctx, args, stream, repoData.AsList(), repoData.IsIndexed())
-		}
+// WithContext uses ctx when running a search over repos using for a searchRepos job.
+func (s *searchRepos) WithContext(ctx context.Context) func() error {
```
Comment (camdencheek) [inline internal/search/unindexed/structural.go:53]: nit: `WithContext()` is used in stdlib to mutate the receiver and add the context. I think either following that pattern or using a different name might be less confusing to my brain which is used to seeing it used that way 😄

**Review by camdencheek — APPROVED** [review]

**Review by keegancsmith — APPROVED** [review]

**Comment by rvantonder** [comment]
Screen Shot 2021-08-18 at 1 32 49 PM

Let it be known that I think Go-style indirection sucks. I would never have caught this without integration tests breaking.

**Comment by keegancsmith** [comment]
classic go footgun. Another way to avoid this issue is something like `j := j`. This is why in general I like pointers in go for structs.

**Comment by rvantonder** [comment]
TIL, useful!

**Pull request merged**